pkg/scheduler/algorithm/priorities/even_pods_spread.go

```diff
@@ -15,3 +15,196 @@ limitations under the License.
 */
 
 package priorities
+
+import (
+	"context"
+	"sync"
+	"sync/atomic"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm/predicates"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulernodeinfo "k8s.io/kubernetes/pkg/scheduler/nodeinfo"
+
+	"k8s.io/klog"
+)
+
+type topologyPair struct {
+	key   string
+	value string
+}
+
+type topologySpreadConstrantsMap struct {
+	// The first error that we faced.
+	firstError error
+	sync.Mutex
+
+	// counts store the mapping from node name to so-far computed score of
+	// the node.
+	counts map[string]*int64
+	// total number of matching pods on each qualified <topologyKey:value> pair
+	total int64
+	// topologyPairToNodeNames store the mapping from potential <topologyKey:value>
+	// pair to node names
+	topologyPairToNodeNames map[topologyPair][]string
+}
+
+func newTopologySpreadConstrantsMap(len int) *topologySpreadConstrantsMap {
+	return &topologySpreadConstrantsMap{
+		counts:                  make(map[string]*int64, len),
+		topologyPairToNodeNames: make(map[topologyPair][]string),
+	}
+}
+
+func (t *topologySpreadConstrantsMap) setError(err error) {
+	t.Lock()
+	if t.firstError == nil {
+		t.firstError = err
+	}
+	t.Unlock()
+}
+
+func (t *topologySpreadConstrantsMap) initialize(pod *v1.Pod, nodes []*v1.Node) {
+	constraints := getSoftTopologySpreadConstraints(pod)
+	for _, node := range nodes {
+		labelSet := labels.Set(node.Labels)
+		allMatch := true
+		var pairs []topologyPair
+		for _, constraint := range constraints {
+			tpKey := constraint.TopologyKey
+			if !labelSet.Has(tpKey) {
+				allMatch = false
+				break
+			}
+			pairs = append(pairs, topologyPair{key: tpKey, value: node.Labels[tpKey]})
+		}
+		if allMatch {
+			for _, pair := range pairs {
+				t.topologyPairToNodeNames[pair] = append(t.topologyPairToNodeNames[pair], node.Name)
+			}
+			t.counts[node.Name] = new(int64)
+		}
+		// for those nodes which don't have all required topologyKeys present, it's intentional to
+		// leave counts[nodeName] as nil, so that we're able to score these nodes to 0 afterwards
+	}
+}
+
+// CalculateEvenPodsSpreadPriority computes a score by checking through the topologySpreadConstraints
+// that are with WhenUnsatifiable=ScheduleAnyway (a.k.a soft constraint).
+// For each node (not only "filtered" nodes by Predicates), it adds the number of matching pods
+// (all topologySpreadConstraints must be satified) as a "weight" to any "filtered" node
+// which has the <topologyKey:value> pair present.
```

Comment:
I think this comment could start with a more high level (and more readable) sentence. What we are trying to say is that, for each node, we want to sum up the number of pods in all the nodes in its same topology domain. But also mention that we reverse that number? (total-counts[node])